[incubator/mysqlha] Use volumeClaimTemplates.spec.storageClassName

holmsten opened the pull request:
The `volume.alpha.kubernetes.io/storage-class` annotation seems to be removed(?) in Kubernetes 1.9.

After changing the template to use `volumeClaimTemplates.spec.storageClassName`, the defined storage class is correctly used and volumes are provisioned as expected.

holmsten committed:
Use volumeClaimTemplates.spec.storageClassName

Instead of using volume.alpha storage class annotation which is
removed(?) in 1.9

holmsten committed:
Bump chart version to 0.1.2

holmsten commented:
/assign @foxish

unguiculus commented:
Thanks for updating this. Please follow this pattern: https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#persistence

unguiculus commented:
/assign

holmsten committed:
Undefine persistence.storageClass by default

Add example for persistence.storageClass option in values.yaml to better
comply with chart review guidelines.

holmsten committed:
Merge branch 'master' into mysqlha-storageclass-spec

holmsten committed:
mysqlha: Add appVersion to Chart.yaml

appVersion is now a required Chart metadata field, add mysql version to
this field.

unguiculus commented:
Please have a look at the PVC example again and update accordingly.

holmsten commented:
I assume you want to address:

> PVCs should support specifying an existing claim

Is it even possible to use existing PVCs in a StatefulSet? Would something like this be acceptable?

`values.yaml`:
```
persistence:
  ## Allow using existing volumes specified by label selector 
  ## ref: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#selector
  ## example:
  ## 
  ## selector:
  ##   matchLabels:
  ##     foo: bar
  selector: {}
```
`templates/statefulset.yaml`
```
  volumeClaimTemplates:
  {{- if .Values.persistence.enabled }}
  - metadata:
      name: data
    spec:
      accessModes:
      - {{ .Values.persistence.accessMode | quote }}
      {{- if .Values.persistence.storageClass }}
      storageClassName: {{ .Values.persistence.storageClass }}
      {{- end }}
      resources:
        requests:
          storage: {{ .Values.persistence.size | quote }}
      selector:
{{ toYaml .Values.persistence.selector | indent 8 }}
  {{- end }}
```

unguiculus commented:
Sorry, maybe I wasn't clear. I'd like you to apply the PVC example to the `volumeClaimTemplate`.

Here's an example you could use as a reference: 
https://github.com/kubernetes/charts/blob/master/stable/mongodb-replicaset/templates/mongodb-statefulset.yaml#L225-L250

holmsten committed:
Merge branch 'master' into mysqlha-storageclass-spec

holmsten committed:
[incubator/mysqlha] Change templating for volumeClaimTemplates

* (Re)add annotations templating

* Change accessModes templating to accept list
  Rename persistence.accessMode to persistence.accessModes to
  reflect this change

* If set persistence.storageClass is set to `-`,
  set storageClassName: "", which disables dynamic provisioning

holmsten commented:
I've used the mongodb example as reference, is it up to standards now?

unguiculus commented on incubator/mysqlha/templates/statefulset.yaml:
```diff
@@ -223,21 +223,30 @@ spec:
       - name: scripts
         emptyDir: {}
-      {{- if eq .Values.persistence.enabled false }}
+{{- if not .Values.persistence.enabled }}
```
Don't use `not` and swap the contents of the if/else blocks.

holmsten replied:
Why should not `not` be used?

holmsten replied:
Now swapped

unguiculus commented on incubator/mysqlha/Chart.yaml:
```diff
@@ -1,5 +1,5 @@
 name: mysqlha
-version: 0.1.3
+version: 0.1.4
```
Make it a 0.2.0.

holmsten replied:
Done

unguiculus commented:
cc @jpoon

jpoon commented:
as storage classes were introduced in 1.6, can you update the readme to denote that k8s 1.6 is now required?

aside from that lgtm.

jpoon commented on incubator/mysqlha/values.yaml:
```diff
@@ -34,10 +34,17 @@ mysqlha:
+  ## If set to "-", storageClassName: "", which disables dynamic provisioning
+  ## If undefined (the default) or set to null, no storageClassName spec is
+  ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
+  ##   GKE, AWS & OpenStack)
```
But...what about Azure?!

holmsten replied:
I've never used Azure so I don't know what is used there. `azure-disk` perhaps - referring to https://docs.microsoft.com/en-us/azure/aks/azure-disks-dynamic-pv?

holmsten replied:
Added documentation for default provisioner used in Azure.

holmsten committed:
[incubator/mysqlha] Swap contents of if/else blocks for volumeClaimTemplates

* Denote that k8s 1.6 is now a requirement in readme

* Add documentation for default provisioner used in Azure

* Bump Chart version to 0.2.0

unguiculus commented:
/ok-to-test

unguiculus commented:
/lgtm

unguiculus commented:
/retest
